src/crimson/os/seastore/segment_cleaner.h

```diff
@@ -188,52 +499,78 @@ class SpaceTrackerDetailed : public SpaceTrackerI {
       }
     }
   };
-  const size_t block_size;
-  const size_t segment_size;
+  struct segment_tracker_t {
+    segment_tracker_t(size_t blocks)
+      : segment_map(blocks) {}
+    segment_id_t segment;
+    SegmentMap segment_map;
+  };
 
   // Tracks live space for each segment
-  std::vector<SegmentMap> segment_usage;
+  segment_info_set_t<segment_tracker_t> segment_usage;
 
 public:
-  SpaceTrackerDetailed(segment_id_t num_segments, size_t segment_size, size_t block_size)
-    : block_size(block_size),
-      segment_size(segment_size),
-      segment_usage(num_segments, segment_size / block_size) {}
+  SpaceTrackerDetailed(
+    std::vector<SegmentManager*> sms)
+  {
+    for (auto sm : sms) {
+      if (!sm) {
```

Comment:
Are there any conditions under which a null pointer here is actually valid?

Reply:
yes, ExtentReader::segment_managers may be passed here, and it may contain holes.

Reply:
You need a comment explaining that.  A vector of possibly null pointers is not a normal interface convention.

Reply:
will do:-)